### PR TITLE
Issue #6659: increase mutation coverage for pitest-filters

### PR DESCRIPTION
**Pull request by esilkensen**
The goal of this PR is to increase the mutation coverage for pitest-filters to 100% (#6659).

I will add PR comments explaining my reasoning for some of the changes, and will certainly be happy to iterate on them in response to any feedback.

### CODE REVIEW

File: src/main/java/com/puppycrawl/tools/checkstyle/filters/SuppressWithNearbyCommentFilter.java [hunk 1]
```diff
@@ -501,9 +501,6 @@ private void addTag(String text, int line) {
                 format = CommonUtil.fillTemplateWithStringsByRegexp(
                         filter.influenceFormat, text, filter.commentFormat);
 
-                if (CommonUtil.startsWithChar(format, '+')) {
```
Comment (esilkensen) [inline src/main/java/com/puppycrawl/tools/checkstyle/filters/SuppressWithNearbyCommentFilter.java]: This is the only change in src/main - I don't think it's necessary to strip a leading "+" here, because this string gets sent to `Integer.parseInt` below which can handle a leading "+" (I updated the `testInfluenceFormat` case to make that explicit in the test)

File: src/test/java/com/puppycrawl/tools/checkstyle/filters/SuppressWithNearbyCommentFilterTest.java [hunk 2]
```diff
@@ -420,10 +420,22 @@ public void testInvalidCheckFormat() throws Exception {
     @Test
     public void testAcceptNullLocalizedMessage() {
         final SuppressWithNearbyCommentFilter filter = new SuppressWithNearbyCommentFilter();
-        final TreeWalkerAuditEvent auditEvent = new TreeWalkerAuditEvent(null, null, null, null);
+        final FileContents contents =
+                new FileContents("filename", "//SUPPRESS CHECKSTYLE ignore");
+        contents.reportSingleLineComment(1, 0);
+        final TreeWalkerAuditEvent auditEvent =
+                new TreeWalkerAuditEvent(contents, null, null, null);
```
Comment (esilkensen) [inline src/test/java/com/puppycrawl/tools/checkstyle/filters/SuppressWithNearbyCommentFilterTest.java]: adding a non-null FileContents here causes the test to fail when the code is mutated to allow the `event.getLocalizedMessage() != null` check to go through

File: src/test/java/com/puppycrawl/tools/checkstyle/filters/SuppressWithPlainTextCommentFilterTest.java [hunk 3]
```diff
@@ -289,6 +289,37 @@ public void testInvalidMessageFormat() throws Exception {
         }
     }
 
+    @Test
+    public void testInvalidMessageFormatWithoutOff() throws Exception {
```
Comment (esilkensen) [inline src/test/java/com/puppycrawl/tools/checkstyle/filters/SuppressWithPlainTextCommentFilterTest.java]: this test covers an edge case where the `messageFormat` might be invalid but is not validated if there is only an "on" comment

File: src/test/java/com/puppycrawl/tools/checkstyle/filters/SuppressionCommentFilterTest.java [hunk 4]
```diff
@@ -310,12 +311,45 @@ public void testEqualsAndHashCodeOfTagClass() {
...
+        filter2.setMessageFormat(".*");
+        final Object tag2 =
+                getTagsAfterExecution(filter2, "filename", "//CHECKSTYLE:ON").get(0);
+        assertEquals("Invalid toString result",
+            "Tag[text='CHECKSTYLE:ON', line=1, column=0, type=ON,"
+                + " tagCheckRegexp=.*, tagMessageRegexp=.*]", tag2.toString());
+    }
+
+    @Test
+    public void testCompareToOfTagClass() {
```
Comment (esilkensen) [inline src/test/java/com/puppycrawl/tools/checkstyle/filters/SuppressionCommentFilterTest.java]: for complete coverage of `compareTo`, we need to compare two tags on the same line

File: src/test/java/com/puppycrawl/tools/checkstyle/filters/SuppressionCommentFilterTest.java [hunk 5]
```diff
@@ -472,12 +518,17 @@ public void testTagsAreClearedEachRun() {
...
      */
-    private static List<?> getTagsAfterExecution(SuppressionCommentFilter filter,
+    private static List<Comparable<Object>> getTagsAfterExecution(SuppressionCommentFilter filter,
             String filename, String... lines) {
         final FileContents contents = new FileContents(filename, lines);
-        contents.reportSingleLineComment(1, 0);
+        for (int lineNo = 1; lineNo <= lines.length; lineNo++) {
+            final int colNo = lines[lineNo - 1].indexOf("//");
+            if (colNo >= 0) {
+                contents.reportSingleLineComment(lineNo, colNo);
```
Comment (esilkensen) [inline src/test/java/com/puppycrawl/tools/checkstyle/filters/SuppressionCommentFilterTest.java]: updated this test helper to try and automatically extract single line comments from the `lines` input, to be able to report multiple comments

File: src/test/java/com/puppycrawl/tools/checkstyle/filters/SuppressionCommentFilterTest.java [hunk 6]
```diff
@@ -472,12 +518,17 @@ public void testTagsAreClearedEachRun() {
...
+            }
+        }
         final TreeWalkerAuditEvent dummyEvent = new TreeWalkerAuditEvent(contents, filename,
-                new LocalizedMessage(1, null, null, null, null, Object.class, null), null);
+                new LocalizedMessage(1, null, null, null, null, Object.class, ""), null);
```
Comment (esilkensen) [inline src/test/java/com/puppycrawl/tools/checkstyle/filters/SuppressionCommentFilterTest.java]: changing the message from `null` to `""` here avoids a NPE when passing `event.getMessage()` to the matcher - alternatively, would it be better for that code to check for a null message itself?

File: src/test/java/com/puppycrawl/tools/checkstyle/filters/XpathFilterElementTest.java [hunk 7]
```diff
@@ -63,7 +64,7 @@ protected String getPackageLocation() {
     public void testMatching() throws Exception {
         final String xpath = "/CLASS_DEF[./IDENT[@text='InputXpathFilterElementSuppressByXpath']]";
         final XpathFilterElement filter = new XpathFilterElement(
-                "InputXpathFilterElementSuppressByXpath", "Test", null, null, xpath);
+                (String) null, null, null, null, xpath);
```
Comment (esilkensen) [inline src/test/java/com/puppycrawl/tools/checkstyle/filters/XpathFilterElementTest.java]: updated the tests throughout this file to try and isolate the parts that are being tested, e.g. the xpath expression here. I thought it might make the tests clearer, but not all of those changes are necessary for full coverage - happy to revert/adjust

**Comment by romani** [comment]
@esilkensen , please rebase to resolve conflict.

File: src/test/java/com/puppycrawl/tools/checkstyle/filters/SuppressionCommentFilterTest.java (hunk 5 above)
Comment (pbludov) [reply src/test/java/com/puppycrawl/tools/checkstyle/filters/SuppressionCommentFilterTest.java]: It will be more natural to iterate over `lines` from `0` to `length`. And call `reportSingleLineComment(lineNo + 1, colNo)`.

File: src/test/java/com/puppycrawl/tools/checkstyle/filters/SuppressionCommentFilterTest.java (hunk 6 above)
Comment (pbludov) [reply src/test/java/com/puppycrawl/tools/checkstyle/filters/SuppressionCommentFilterTest.java]: In the real world, the `key` parameter is always specified, so `customMessage` is optional. But in testing the `key` parameter sometimes is omitted, causing the NPE if the `customMessage` is missing too. I like this change.

File: src/test/resources/com/puppycrawl/tools/checkstyle/filters/suppresswithplaintextcommentfilter/InputSuppressWithPlainTextCommentFilterWithCustomOnComment.java [hunk 8]
```diff
@@ -0,0 +1,6 @@
+package com.puppycrawl.tools.checkstyle.filters.suppresswithplaintextcommentfilter;
+
+public class InputSuppressWithPlainTextCommentFilterWithCustomOnComment {
+    // cs-on
+    //	has tab here
```
Comment (pbludov) [inline src/test/resources/com/puppycrawl/tools/checkstyle/filters/suppresswithplaintextcommentfilter/InputSuppressWithPlainTextCommentFilterWithCustomOnComment.java]: Does it really need a tab character? We have an issue for this: https://github.com/checkstyle/checkstyle/issues/5941

Comment (esilkensen) [reply src/test/resources/com/puppycrawl/tools/checkstyle/filters/suppresswithplaintextcommentfilter/InputSuppressWithPlainTextCommentFilterWithCustomOnComment.java]: I added a SQL file instead with the example "COUNT(*)" check from the docs

File: src/test/java/com/puppycrawl/tools/checkstyle/filters/SuppressionCommentFilterTest.java [hunk 9]
```diff
@@ -310,12 +311,45 @@ public void testEqualsAndHashCodeOfTagClass() {
...
-        final Object tag =
-                getTagsAfterExecution(filter, "filename", "//CHECKSTYLE:OFF").get(0);
+        final SuppressionCommentFilter filter1 = new SuppressionCommentFilter();
+        final Object tag1 =
+                getTagsAfterExecution(filter1, "filename", "//CHECKSTYLE:OFF").get(0);
         assertEquals("Invalid toString result",
             "Tag[text='CHECKSTYLE:OFF', line=1, column=0, type=OFF,"
-                    + " tagCheckRegexp=.*, tagMessageRegexp=null]", tag.toString());
+                    + " tagCheckRegexp=.*, tagMessageRegexp=null]", tag1.toString());
+
```
[the remark this answers is not in the document]
Comment (esilkensen) [reply src/test/java/com/puppycrawl/tools/checkstyle/filters/SuppressionCommentFilterTest.java]: Sounds good, split it into a second test case below

File: src/test/java/com/puppycrawl/tools/checkstyle/filters/SuppressWithNearbyCommentFilterTest.java [hunk 10]
```diff
@@ -420,10 +420,22 @@ public void testInvalidCheckFormat() throws Exception {
...
         assertTrue("Filter should accept null localized message", filter.accept(auditEvent));
     }
 
+    @Test
+    public void testAcceptNullFileContents() {
+        final SuppressWithNearbyCommentFilter filter = new SuppressWithNearbyCommentFilter();
+        final TreeWalkerAuditEvent auditEvent = new TreeWalkerAuditEvent(null, null,
```
Comment (romani) [inline src/test/java/com/puppycrawl/tools/checkstyle/filters/SuppressWithNearbyCommentFilterTest.java]: Too much nulls in arguments, please make variable `content = null` and use it in constructor.

Please apply the same for all other test method you changed

File: src/test/java/com/puppycrawl/tools/checkstyle/filters/SuppressWithPlainTextCommentFilterTest.java [hunk 11]
```diff
@@ -290,6 +290,38 @@ public void testInvalidMessageFormat() throws Exception {
...
+        filterCfg.addAttribute("messageFormat", "e[l");
+
+        final DefaultConfiguration checkCfg = createModuleConfig(RegexpSinglelineCheck.class);
+        checkCfg.addAttribute("format", "^.*COUNT\\(\\*\\).*$");
+        checkCfg.addAttribute("message", "Don''t use COUNT(*), use COUNT(1) instead.");
+
+        final String[] suppressed = CommonUtil.EMPTY_STRING_ARRAY;
+
+        final String[] violationMessages = {
+            "2: Don't use COUNT(*), use COUNT(1) instead.",
```
Comment (romani) [inline src/test/java/com/puppycrawl/tools/checkstyle/filters/SuppressWithPlainTextCommentFilterTest.java]: Please move message to variable to reuse in both places

File: src/test/java/com/puppycrawl/tools/checkstyle/filters/SuppressionCommentFilterTest.java [hunk 12]
```diff
@@ -300,24 +301,53 @@ private void verifySuppressed(Configuration moduleConfig, String fileName,
 
     @Test
     public void testEqualsAndHashCodeOfTagClass() {
-        final SuppressionCommentFilter filter = new SuppressionCommentFilter();
-        final Object tag =
-                getTagsAfterExecution(filter, "filename", "//CHECKSTYLE:OFF").get(0);
+        final Object tag = getTagsAfterExecution("//CHECKSTYLE:OFF").get(0);
```
Comment (romani) [inline src/test/java/com/puppycrawl/tools/checkstyle/filters/SuppressionCommentFilterTest.java]: Please name it getTagsAfterExecutionOnDefaulFilter

File: src/test/java/com/puppycrawl/tools/checkstyle/filters/SuppressionXpathSingleFilterTest.java [hunk 13]
```diff
@@ -165,6 +166,30 @@ public void testNonMatchingFileRegexp() throws Exception {
...
 
+    @Test
+    public void testInvalidFileRegexp() {
+        final SuppressionXpathSingleFilter filter = new SuppressionXpathSingleFilter();
+        try {
+            filter.setFiles("e[l");
+            fail("PatternSyntaxException is expected");
+        }
+        catch (PatternSyntaxException ex) {
+            // test passed
```
Comment (romani) [inline src/test/java/com/puppycrawl/tools/checkstyle/filters/SuppressionXpathSingleFilterTest.java]: Please test message of exception

File: src/test/java/com/puppycrawl/tools/checkstyle/filters/XpathFilterElementTest.java [hunk 14]
```diff
@@ -117,6 +118,19 @@ public void testComplexQuery() throws Exception {
         assertFalse("Event should be rejected", filter.accept(eventThree));
     }
 
+    @Test
+    public void testInvalidChecksRegexp() {
+        try {
+            final Object test = new XpathFilterElement(
+                    ".*", "e[l", ".*", "moduleId", "query");
+            fail("Expected was expected but got " + test);
```
Comment (romani) [inline src/test/java/com/puppycrawl/tools/checkstyle/filters/XpathFilterElementTest.java]: Exception is expected

File: src/test/java/com/puppycrawl/tools/checkstyle/filters/XpathFilterElementTest.java [hunk 15]
```diff
@@ -223,7 +248,7 @@ public void testNonMatchingFileNameModuleIdAndCheck() throws Exception {
     public void testNullModuleIdAndNonMatchingChecks() throws Exception {
         final String xpath = "NON_MATCHING_QUERY";
         final XpathFilterElement filter = new XpathFilterElement(
-                "InputXpathFilterElementSuppressByXpath", "NonMatchingRegexp", null, null, xpath);
+                (String) null, null, null, null, xpath);
```
Comment (romani) [inline src/test/java/com/puppycrawl/tools/checkstyle/filters/XpathFilterElementTest.java]: Please move such updates to separate commit of the same PR, as it is not related to issue

**Comment by esilkensen** [comment]
@romani thanks for the review - I have tried to address all comments, please let me know if I missed anything or for any additional changes.

I feel better about XpathFilterElementTest now - I think I was able to reduce my changes there to only those required for pitest coverage.
